enhancements/kube-apiserver/unsupported-file-sync.md

```diff
@@ -0,0 +1,153 @@
+- allow mounting unsupported content as files in the kube-apiserver static pods
+
+### Non-Goals
+
+- allow creating files anywhere on a node
+
+## Proposal
+
+The `openshift-kube-apiserver-operator` namespace gets a new revisioned
+secret `unsupported-kube-apiserver-content` that is going to be used to
+install content inside of the `/etc/kubernetes/unsupported/kube-apiserver` directory.
+This secret will be populated by an observer inside of the cluster-kube-apiserver-operator.
+The observer watches the `unsupported-kube-apiserver-content` secret in
+the `openshift-config` namespace and synchronizes its content to the
+secret inside of the `openshift-kube-apiserver-namespace`.
+
+### Workflow Description
+
+The user adds content to the `data` and `stringData` maps in the
+`openshift-config/unsupported-kube-apiserver-content` secret and
+waits for a rollout of a new kube-apiserver revision that contains their changes.
+
+### API Extensions
+
+No API extensions.
+
+### Implementation Details
+
+#### New kube-apiserver operator observer
+
+A new observer is added to the cluster-kube-apiserver-operator. This observer
+watches the `data` and `stringData` fields of the `openshift-config/unsupported-kube-apiserver-content`
+secrets and synchronizes them to the `openshift-kube-apiserver-operator/unsupported-kube-apiserver-content`
+secret.
+
+If the `data` or `stringData` fields of the source secret are non-empty, the observer
+sets the `UnsupportedFilesUpgradeable: False` condition in the `kube-apiserver`
+clusteroperator.
+
+#### Installing the unsupported content
+
+`openshift-config/unsupported-kube-apiserver-content` is a revisioned and optional
+secret. Its last revision's content gets installed into the `/etc/kubernetes/unsupported/kube-apiserver`
+directory.
+
+In order to accommodate these changes, the [installer command](https://github.com/openshift/cluster-kube-apiserver-operator/blob/ab3736b9773f1b9850fb046226a1ca8b4342cb1c/vendor/github.com/openshift/library-go/pkg/operator/staticpod/installerpod/cmd.go#L85-L86)
+and the [installer controller](https://github.com/openshift/cluster-kube-apiserver-operator/blob/ab3736b9773f1b9850fb046226a1ca8b4342cb1c/vendor/github.com/openshift/library-go/pkg/operator/staticpod/controller/installer/installer_controller.go#L156)
+need to be extended. This functionality should be optional, with an opt-in scheme.
+
+### Risks and Mitigations
+
+**Risk**: somebody breaks their cluster so badly it's no longer recoverable and opens a support case.
+**Mitigation**: support shrugs and closes that case when they notice unsupported config was active.
+
+### Drawbacks
+
+This allows installing random files on control-plane nodes' disks. However, we
+control the directory where such files are installed, and we specifically mark
+any cluster with such a configuration as unsupported.
+
+## Design Details
+
+### Open Questions [optional]
+
+- Should the `*Upgradable: False` condition be reported by a different controller
```

Comment:
imo kube-apiserver-operator should switch to Upgradable=False if this secret exists. We don't know how this configuration affects upgradability and should be defensive here

Reply:
I don't mind either way. But you can still have active pods of a previous revision that still use the unsupported files, which is where this question is aimed.

Reply:
you mean some pods which didn't yet apply a new config are upgradable? Yes, but not the cluster as whole

Reply:
1. you remove the secret
2. you set upgradable true because the secret (or its values don't exist)
3. new kube-apiserver pods' revision starts rolling out
4. an old revision is still active, new revision may fail to rollout
5. the cluster no longer reports upgradable:false since 2., but the unupgradable revision is still active